src/erc721.rs

```diff
@@ -0,0 +1,367 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+use sp_runtime::traits::Hash as HashT;
+use support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchError,
+    dispatch::DispatchResult, ensure, traits::Randomness,
+};
+use system::ensure_signed;
+
+// Encoding library
+use codec::Encode;
+
+pub trait Trait: system::Trait + balances::Trait + contracts::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    // Something that provides randomness generation in the runtime.
+    type Randomness: Randomness<Self::Hash>;
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        // No owner set for token
+        TokenNoOwner,
+
+        // Got an overflow after adding
+        Overflow,
+
+        // Got an underflow after subing
+        Underflow,
+    }
+}
+
+decl_event!(
+    pub enum Event<T> 
+        where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash {
+
+        // Token transfer event
+        Transfer(Option<AccountId>, Option<AccountId>, Hash),
+
+        // One token approved to an account
+        Approval(AccountId, AccountId, Hash),
+
+        // All tokens owned are approved to other account
+        ApprovalForAll(AccountId, AccountId, bool),
+    }
+);
+
+decl_storage! {
+    trait Store for Module<T: Trait> as ERC721 {
+        // Token count owned by an account
+        pub OwnedTokensCount get(balance_of): map T::AccountId => u64;
+
+        // Token owner's account id
+        pub TokenOwner get(owner_of): map T::Hash => Option<T::AccountId>;
+
+        // Token delegate's account id
+        pub TokenApprovals get(get_approved): map T::Hash => Option<T::AccountId>;
+
+        // Account deledate's account id
+        pub OperatorApprovals get(is_approved_for_all): map (T::AccountId, T::AccountId) => bool;
+
+        // Total count of minted token
+        pub TotalSupply get(total_supply): u64;
+
+        // Map of token index to token id
+        pub AllTokens get(token_by_index): map u64 => T::Hash;
+
+        // Last token index
+        pub AllTokensIndex: map T::Hash => u64;
+
+        // Map of account and index to token id
+        pub OwnedTokens get(token_of_owner_by_index): map (T::AccountId, u64) => T::Hash;
```

Comment:
I think current design is efficient for get owned token one by one, enumeration as implemented. Use the vector is other approach,  we can get all tokens owned by an account in one query. But each time, we need get them all even just need one token. I think both solutions have the pros and cons. In real world case, for instance, we show tokens in web. There is paging mechanishm, we just show about 20 token each page. So current solution seems good. What do you take?

Reply:
mhm. I think in terms of inoperability it would be nice if the ERC721 interface looks more a less the same on each chain.  I think if possible, we should return an array with all tokenId's of an account same as on Ethereum.

Reply:
Ok, I can change the storage if you think it is necessary. I'd like to ask a question about the approach. If the leaf storage is a vector with frequent updates, too much state change may bring big history data in storage?

Reply:
@manuelpolzhofer does ERC721 also return an array if there are 2 million NFTs? Or does it support pagination in that case?

Reply:
I'm fine with storing the tokens individually, but you are using a `map`, where I suggest a `double map` so you can iterate much more efficiently.

Reply:
Yes, I will use double map to replace map.